File: generator_web_app/src/modules/api/herdunits.ts
// API methods for managing herd unit objects
// Author: Michael B. Lance

//---------------------------------------------------------------------------------------------------------------------------//

import { HerdUnit, Survey } from '@/types/generatorobjects.ts';
import type { HerdUnitIntf, SurveyIntf } from '@/types/generatorobjects.ts';
import { ApiError } from '@/modules/api/errors.ts'
import { api_url } from '@/modules/api/apiV1Methods.ts';

//---------------------------------------------------------------------------------------------------------------------------//

export async function getHerdUnitSurveys(herd_unit_id: string): Promise<Survey[]> {
    const response = await fetch(`${api_url}/herd-units/${herd_unit_id}/surveys`, {
        method: 'GET',
        credentials: 'include',
        headers: {
            'Content-Type': 'application/json'
        },
    });
    if (!response.ok) throw new ApiError(await response.json());
    const resp = await response.json();
    let surveys = [];
    for (const survey of resp) surveys.push(new Survey(survey as SurveyIntf));
    return surveys;
}

//---------------------------------------------------------------------------------------------------------------------------//

export async function createHerdUnit(project_id: number, name: string): Promise<HerdUnit> {
    const response = await fetch(`${api_url}/herd-units`, {
        method: 'POST',
        credentials: 'include',
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify({
            'project_id': project_id,
            'name': name
        }),
    });
    if (!response.ok) throw new ApiError(await response.json());
    
    return new HerdUnit(await response.json() as HerdUnitIntf);
}

//---------------------------------------------------------------------------------------------------------------------------//

File: annotation_api/app/routers/herdunits/herdunits.py
# Endpoints for managing herd units in the API 
# Author: Michael B. Lance

#---------------------------------------------------------------------------------------------------------------------------#

from uuid import UUID

from flask import Blueprint, abort
from flask_login import login_required
from flask_pydantic import validate
from psycopg.errors import DatabaseError

from app.extensions import base
from database import ObjectNotFound


from .herdunit_validators import CreateHerdUnit

herdunitBp = Blueprint('herd_units', __name__, url_prefix='/api/v1/herd-units')

#---------------------------------------------------------------------------------------------------------------------------#
# GET 

@herdunitBp.get('/all')
@login_required
def get_all():
	'''

	'''
	return ''

#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

@herdunitBp.get('/<string:herd_unit_id>')
@login_required
def get_by_id(herd_unit_id: str):
	'''

	'''
	herd_unit = base.get_herd_unit(UUID(herd_unit_id))

	if herd_unit is None:
		abort(404, f'Herd Unit with ID {herd_unit_id} was not found!')
		
	else:
		return herd_unit.to_dict()


#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~#

@herdunitBp.get('/<string:herd_unit_id>/surveys')
@login_required
def get_surveys(herd_unit_id: str):
	'''
	Retrieve all surveys associated with a herd unit.
	---
	parameters:
		- name: survye_id
		in: path
		type: string
		required: true
	responses:
		200:
			description: List of surveys.
		400:
			description: Invalid UUID format.
		404:
			description: No surveys found.
		500:
			description: Database error.
	'''
	try:
		surveys = base.get_herd_unit_surveys(UUID(herd_unit_id))

	except ValueError as e:
		abort(400, str(e))
	except ObjectNotFound as e:
		abort(404, str(e))
	except (DatabaseError, Exception) as e:
		abort(500)

	return [survey.to_dict() for survey in surveys], 200

#---------------------------------------------------------------------------------------------------------------------------#
# POST

@herdunitBp.post('')
@login_required
@validate()
def create(body: CreateHerdUnit):
	'''

	'''
	try:
		herd_unit = base.create_herd_unit(body.model_dump())
	except Exception as e:
		print(e)
		abort(500)

	return herd_unit.to_dict(), 201